docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+## Goals
+
+Safety is a top priority. The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators).
+
+C++ interoperability aims to provide a clear, well-defined mapping for APIs that can be imported, and to draw a clear boundary between these APIs and APIs that cannot be imported. A user should be able to read the C++ interoperability documentation and have a good idea of both how much of their C++ API will be able to be imported in Swift and what that API will look like once it is imported. 
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
+
+In this section, the roadmap will discuss several of the most common C++ API patterns and how they might be imported by the Swift compiler using the goals outlined above.
+
+### Definitions
```

Comment:
These definitions feel verbose after reading the proceeding section. I think we can probably just delete the whole thing. Is there any useful information here?

Reply:
I personally appreciated it -- as someone with less understanding of these things

Reply:
Yeah, let's move them to the top of the document and define a few other things. Specifically for owned types, I want to remove the specifics about special members (that will be covered in detail below).

Reply:
The title “definitions” makes it sound like this is a collection of unrelated terminology, but I believe this is really the list of “common C++ API patterns” you were talking about before. If so, you may want to retitle this section.

I also think there’s a sort of hierarchy that you might want to represent more clearly here by reorganizing and referencing back to previous definitions:

* Reference types
* Owned types
    * Trivial types (a specific kind of owned type)
* Projections
    * Iterators (a specific kind of projection)

Reply:
IMO these terms are not universally understood, cannot be assumed, and *must* be defined up-front.

Reply:
I agree we should keep these. I still have a difficult time keeping them straight without some examples (even after reading this a couple times).

Reply:
I agree, these definitions should be kept in. It really helped my reading of the document.